### PR TITLE
[CORE-2256] Fix hint return and display

### PR DESCRIPTION
**Pull request by robert-uhl**
Reworked paging and hints. Added a plethora of tests against a testing Loki instance. Worked with @pleeko to test expected vs. actual results.

### CODE REVIEW

File: src/server/logs/cmds/cmds.go [hunk 1]
```diff
@@ -7,18 +7,20 @@ import (
 	"os"
 	"time"
 
+	"github.com/alessio/shellescape"
```
Comment (jrockway) [inline src/server/logs/cmds/cmds.go:10]: Good idea.

File: src/internal/lokiutil/testloki/testloki.go [hunk 2]
```diff
@@ -32,13 +32,25 @@ type TestLoki struct {
...
+func WithoutOldSampleRejection(config *(map[string]any)) {
+	(*config)["limits_config"].(map[string]any)["reject_old_samples"] = false
+}
+
+func WithCreationGracePeriod(d time.Duration) func(config *(map[string]any)) {
+	return func(config *(map[string]any)) {
+		(*config)["limits_config"].(map[string]any)["creation_grace_period"] = d
+	}
+}
+
```
Comment (jrockway) [inline src/internal/lokiutil/testloki/testloki.go:46]: I would just set these to infinite in the default config.  That was what I had in mind when making the config file.

**Review by jrockway — APPROVED** [review]
As explained on Slack, I don't think we need this feature.  The chances of having a lot of lines with the same timestamp is pretty much zero, and even then, you still can't get duplicate-timestamped lines that exceed the loki server's limit.  I'd just let it be.

You should keep the changes that let you ingest old samples.  I'd expect some existing tests to be failing if they can't, though.

**Comment by pleeko** [comment]
TY for implementing these fixes!

**Review by emmajsadams — APPROVED** [review]

**Comment by robert-uhl** [comment]
>  The chances of having a lot of lines with the same timestamp is pretty much zero, and even then, you still can't get duplicate-timestamped lines that exceed the loki server's limit.

With just `promtail`, you’re correct that the odds of it happening by chance are zero or close to it, but anyone in the cluster is able to write logs with any timestamp (within the bounds Loki permits). I think that it’s necessary to be at least aware enough of the possibility to not crash or loop infinitely when encountering runs of logs with the same timestamp, and the complexity involved even in handling that is enough that it gets one pretty close to just handling duplicate timestamps generally.

> even then, you still can't get duplicate-timestamped lines that exceed the loki server's limit

This is also true; it’s a limitation of Loki which cannot be worked around (I’m pretty sure).

I share your concern regarding the amount of complexity this adds, but I think that it’s necessary, and hope that the extensive test coverage provides some level of protection from implementation bugs. It’d be great if Loki guaranteed unique timestamps, or if we didn’t rely on it, but I believe that this is the least bad solution.

**Pull request merged**